src/compute-client/src/controller.rs

```diff
@@ -452,6 +455,17 @@ where
         }
     }
 
+    /// Update compute configuration.
+    pub fn update_configuration(&mut self, config_params: BTreeSet<ComputeParameter>) {
```

Comment:
Given the logic in #17094, perhaps it would be more consistent to employ a `BTreeMap` to represent `config_params`? It is also a bit unclear to me what the semantics of `UpdateConfiguration( { max_result_size = 10, max_result_size = 100} )` is exactly, which is possible to represent if `UpdateConfiguration` has a `BTreeSet`?

Reply:
You are right, representing them as a `BTreeSet` doesn't make much sense. I guess I didn't think things through properly when I introduced that. The intent was to ensure that there is only ever one of each parameter, but clearly this is not ensured at all! We could make it work by having a custom `Eq` impl for `ComputeParameter` that simply ignores the value, but that would be an ugly hack.

A `BTreeMap` makes things better, at the expense of static type checking. I think what we should have instead is a `ComputeParameters` struct with all fields being `Option`s, and pass that around. That's the most straightforward solution too. I initially decided against it, because it seemed expensive to update individual parameters that way, in particular since they are transmitted through the network. But a) adapter always passes all parameters to the controllers anyway on every config updated, b) config updates are relatively infrequent events and c) protobuf actually has an efficient encoding for sparse messages. So there is really no reason at the moment to not use a config struct.

> It is also a bit unclear to me what the semantics of UpdateConfiguration( { max_result_size = 10, max_result_size = 100} ) is exactly, which is possible to represent if UpdateConfiguration has a BTreeSet?

Eventually, we iterate through the set and either deduplicate or apply the parameters in order. The one with the larger value would be seen last and would stick, is my guess.